os/net/lwip/test/unit/mbox/mbox_shared.c

```diff
@@ -0,0 +1,182 @@
+static sys_mbox_t g_tcpmbox;
+static sem_t *g_signal_atoc = NULL;
+static sem_t *g_signal_ctoa = NULL;
+static sem_t *g_signal_btoc = NULL;
+static sem_t *g_signal_ctob = NULL;
+
+void mbox_shared_event1(void) {
+  printf("[EVT] let task C know mbox is empty \t%s:%d\n", __FUNCTION__,
+         __LINE__);
+  sem_post(g_signal_atoc);
+}
+
+void mbox_shared_event2(void) {
+  printf(
+    "[EVT] task b post signal to task a that mbox is not empty \t%s:%d\n",
+    __FUNCTION__, __LINE__);
+}
+
+void mbox_shared_event3(uint32_t front, uint32_t rear) {
+  if (front == rear) {
+    printf(
+      "[EVT] task b exit sem_wait even though mbox is full(%u/%u) because "
+      "it "
+      "receives signal that task b sent(not emtpy) \t%s:%d\n",
+      front, rear, __FUNCTION__, __LINE__);
+    assert(0);
+  } else {
+    printf(
+      "[EVT] queue is not full(%u/%u) so it's ok to exit sem_wait "
+      "\t%s:%d\n",
+      front, rear, __FUNCTION__, __LINE__);
+  }
+} 
+
+static void *control_thread(void *arg) {
+  /*
+  1) Task A called mbox_fetch. But there is no data in mbox, task A waits an
+  event with mbox->mail
+  2) Task B called mbox post. And call
+  sys_arch_sem_signal(mbox->mail) to signal Task A
+  3) If context switching
+  didn't happen then Task B is doing mbox post repeatedly.
+  4) mbox can be fulled
+  because Task B posts data again and again. Then it calls sys_arch_sem_wait()
+  to get a signal that mbox is not full
+  5) sys_arch_sem_wait() from Task B return immediately because task A didn't
+  receive a signal.
+  6) Unexpected behavior will happen.
+  */
+
+  /*  send signal to thread a to call sys_arch_mbox_fetch */
+  printf("[COND] run task A \t%s:%d\n",  __FUNCTION__, __LINE__);
+  sem_post(g_signal_ctoa);
+  /*  wait event that task a wait event with mbox->mail because it's empty */
+  printf("[COND] wait until task A fetch event \t%s:%d\n",  __FUNCTION__, __LINE__);
+  sem_wait(g_signal_atoc);
+  printf("[COND] task A wait not empty signal \t%s:%d\n",  __FUNCTION__, __LINE__);
+  /*  let task b to post data until g_tcpmbox is full */
+  printf("[COND] run task B to post data \t%s:%d\n",  __FUNCTION__, __LINE__);
+  sem_post(g_signal_ctob);
+
+  return NULL;
+}
+
+static void *a_thread(void *arg) {
+  uint32_t recv_cnt = 0;
+  printf("[taskA] wait control signal \t%s:%d\n", __FUNCTION__, __LINE__);
+  sem_wait(g_signal_ctoa);
+  while (recv_cnt < QUEUE_MSG_CNT) {
+    lwip_msg_s *msg = NULL;
+    printf("[taskA] -->sys_arch_mbox_fetch \t%s:%d\n", __FUNCTION__,
+           __LINE__);
+    uint32_t res = sys_arch_mbox_fetch(&g_tcpmbox, (void **)&msg, 0);
+    if (res != 0) {
+      assert(0);
+    }
+    free(msg);
+    recv_cnt++;
+  }
+  return NULL;
+}
+
+static void *b_thread(void *arg) {
+  printf("[taskB] wait control signal  \t%s:%d\n", __FUNCTION__, __LINE__);
+  sem_wait(g_signal_ctob);
+  uint32_t send_cnt = 0;
+  while (send_cnt < QUEUE_MSG_CNT) {
+    lwip_msg_s *msg = (lwip_msg_s *)malloc(sizeof(lwip_msg_s));
```

Comment:
I think it is kernel side, so you should use `kmm_malloc` instead of `malloc`.

Reply:
Actually this test case is tested in linux. I'll modify it with wrapper.